ride/swap.ride

```diff
@@ -0,0 +1,277 @@
+  }
+}
+
+func dataMappingPoolAssets(internalAmountAssetStr: Int, internalPriceAssetStr: Int) = {
+  makeString(
+    ["%d%d",
+    internalAmountAssetStr.toString(),
+    internalPriceAssetStr.toString()],
+    SEP)
+}
+
+func managerPublicKeyOrUnit() = match keyManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func pendingManagerPublicKeyOrUnit() = match keyPendingManagerPublicKey().getString() {
+  case s: String => s.fromBase58String()
+  case _: Unit => unit
+}
+
+func isManager(i: Invocation) = {
+  match managerPublicKeyOrUnit() {
+    case pk: ByteVector => i.callerPublicKey == pk
+    case _: Unit => i.caller == this
+  }
+}
+
+func mustManager(i: Invocation) = {
+  i.isManager() || "permission denied".throw()
+}
+
+func getAccBalance(assetId: String) = {
+  if(assetId == "WAVES") then wavesBalance(this).available else assetBalance(this, fromBase58String(assetId))
+}
+
+@Callable(i)
+func swapCalculateREADONLY(
+  amountIn: Int,
+  assetIn: String,
+  assetOut: String
+) = {
+  let prFee = getInteger(protocolFee()).value()
+  let feeProtocolAmountCalc = fraction(amountIn.toBigInt(), prFee.toBigInt(), feeScale)
+  let plFee = getInteger(poolFee()).value()
+  let feePoolAmountCalc = fraction(amountIn.toBigInt(), plFee.toBigInt(), feeScale)
+
+  let feeProtocolAmount = if (feeProtocolAmountCalc.toInt() == 0) then {
+    toBigInt(1)
+  } else {
+    feeProtocolAmountCalc
+  }
+  let feePoolAmount = if (feePoolAmountCalc.toInt() == 0) then {
+    toBigInt(1)
+  } else {
+    feePoolAmountCalc
+  }
+  let cleanAmountIn = amountIn.toBigInt() - feeProtocolAmount - feePoolAmountCalc
+
+  let lpNonReverse = invoke(
+    factoryContract,
+    "getLpAssetFromPoolAssetsREADONLY",
+    [assetIn, assetOut],
+    []
+  )
+
+  let lpReverse = invoke(
+    factoryContract,
+    "getLpAssetFromPoolAssetsREADONLY",
+    [assetOut, assetIn],
+    []
+  )
+
+  let (isReverse, pool) = if (lpNonReverse != unit) then {
+    let isReverse = false
+    let pool = invoke(
+      factoryContract,
+      "getPoolAddressFromLpAssetREADONLY",
+      [lpNonReverse],
+      []
+    ).exactAs[String]
+    (isReverse, pool)
+  } else if (lpReverse != unit) then {
+    let isReverse = true
+    let pool = invoke(
+      factoryContract,
+      "getPoolAddressFromLpAssetREADONLY",
+      [lpReverse],
+      []
+    ).exactAs[String]
+    (isReverse, pool)
+  } else {
+    poolNotExist
+  }
+
+  let poolStatus = invoke(
+    factoryContract,
+    "getPoolStatusREADONLY", 
+    [pool], 
+    []
+  ).exactAs[Int]
+  strict check = poolStatus == 1 || poolDisabled
+  let poolAdr = addressFromStringValue(pool.valueOrElse(""))
+
+  let res = invoke(
+    poolAdr,
+    "calculateAmountOutForSwapREADONLY",
+    [cleanAmountIn.toInt(), isReverse],
+    []
+  ).exactAs[Int]
+
+
+  (nil,
+  res)
+}
+
+@Callable(i)
+func swap(
+  amountOutMin: Int,
+  assetOut: String,
+  addressTo: String 
```

Comment:
What is the point for `addressTo` in the top interface function? Should a user every time check this when he sign a transaction.

Does not look safe for a user. And does not a clear utility.

Reply:
I think it's pretty clear. uniswap/curve also use it. And it is pretty flexible solution

Reply:
On ethereum-like chains maybe. Here you are the first with that nd it looks pretty ugly.

At least you should make another function without addressTo.